ntoskrnl/se/sid.c

```diff
@@ -598,7 +667,32 @@ SeCaptureSidAndAttributesArray(
                 SidAndAttributes[i].Sid = (PSID)CurrentDest;
                 RtlCopyMemory(CurrentDest, SrcSidAndAttributes[i].Sid, SidLength);
 
-                /* Sanity checks */
+                /*
+                 * Now it's time to do validation checks. As
+                 * we're copying the SIDs into the final array
+                 * buffer, each time we do that we must validate
+                 * each condition, in following order:
+                 *
+                 * 1 -- The subauthority count of the SID hasn't
+                 *      changed.
+                 *
+                 * 2 -- The length of the SID hasn't changed, and
+                 *      it's literally the same like before.
+                 *
+                 * 3 -- The length size of the SID is equal
+                 *      on itself.
+                 *
+                 * 4 -- The SID meets the requirements imposed
+                 *      by the security manager that it's valid.
+                 *
+                 * Any assertion violation of these conditions means
+                 * that there's been an attempt on changing the SID
+                 * properties whilst the function was copying the
+                 * SIDs into the final buffer. This can bring a lot
+                 * of issues so at this point just bail out.
+                 */
+                ASSERT(ValidateArray[i].SubAuthorityCount == Sid->SubAuthorityCount);
+                ASSERT(RtlLengthSid(SidAndAttributes[i].Sid) == ValidateArray[i].SidLength);
```

Comment:
But shouldn't we do these checks live on release builds as well, and fail properly in case something bad happens?